src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs

```diff
@@ -206,5 +291,160 @@ public Task UnloadAsync()
 
             return Task.CompletedTask;
         }
+
+        // NuGet calls this method to wait project to nominate restoring.
+        // If the project has no pending restore data, it will return a completed task.
+        // Otherwise a task which will be completed once the project nominate the next restore
+        // the task will be cancelled, if the project system decide it no longer need restore (for example: the restore state has no change)
+        // the task will be failed, if the project system runs into a problem, so it cannot get correct data to nominate a restore (DT build failed)
+        public Task WhenNominated(CancellationToken cancellationToken)
+        {
+            lock (SyncObject)
+            {
+                if (!CheckIfHasPendingNomination() || cancellationToken.IsCancellationRequested)
+                {
+                    return Task.CompletedTask;
+                }
+
+                _whenNominatedTask ??= new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+                if (!_wasSourceBlockContinuationSet)
+                {
+                    _wasSourceBlockContinuationSet = true;
+
+                    _ = SourceBlock.Completion.ContinueWith(t =>
+                    {
+                        lock (SyncObject)
+                        {
+                            if (_whenNominatedTask is not null)
+                            {
+                                if (t.IsFaulted)
+                                {
+                                    _whenNominatedTask.SetException(t.Exception);
+                                }
+                                else
+                                {
+                                    _whenNominatedTask.TrySetCanceled();
+                                }
+                            }
+                        }
+                    }, CancellationToken.None, TaskContinuationOptions.OnlyOnFaulted, TaskScheduler.Default);
+                }
+            }
+
+            return _whenNominatedTask.Task.WithCancellation(cancellationToken);
+        }
+
+        // True means the project system plans to call NominateProjectAsync in the future.
+        bool IVsProjectRestoreInfoSource.HasPendingNomination
+        {
+            get => CheckIfHasPendingNomination();
+        }
+
+        private bool CheckIfHasPendingNomination()
+        {
+            lock (SyncObject)
+            {
+                Assumes.Present(_project.Services.ActiveConfiguredProjectProvider);
+                Assumes.Present(_project.Services.ActiveConfiguredProjectProvider.ActiveConfiguredProject);
+
+                // Nuget should not wait for projects that failed DTB
+                if (SourceBlock.Completion.IsFaulted || SourceBlock.Completion.IsCompleted)
```

Comment:
also return false when this._enabled is false.  In case the feature is disabled for a reason.